src/pypromice/process/join_l3.py

```diff
@@ -0,0 +1,282 @@
+    #                    df.loc[df.HS_combined.notnull(),'HS_combined'], 1)
+    # fit_fn = np.poly1d(fit)
+    
+    # df['HS_combined'] = df['HS_combined'].values \
+    #     - fit_fn(
+    #         df_in.loc[[df_in.z_surf_combined.first_valid_index()],:].index.astype('int64')[0]
+    #     )  +  df_in.loc[df_in.z_surf_combined.first_valid_index(), 'z_surf_combined']
+    # return l3m
+    
+def build_station_dict(config_folder):
+    station_dict = {}
+    
+    # Iterate through all files in the given folder
+    for filename in os.listdir(config_folder):
+        if filename.endswith(".toml"):
+            # Construct full file path
+            file_path = os.path.join(config_folder, filename)
+            
+            # Load TOML file
+            with open(file_path, 'r') as file:
+                data = toml.load(file)
+                station_key = next(iter(data))
+                station_data = data[station_key]
+               
+                if station_data["station_site"] in station_dict:
+                    station_dict[station_data["station_site"]].append(station_key)
+                else:
+                    station_dict[station_data["station_site"]] = [station_key]
+    
+    return station_dict
+    
+def join_l3():
+    args = parse_arguments_joinl3()
+    logging.basicConfig(
+        format="%(asctime)s; %(levelname)s; %(name)s; %(message)s",
+        level=logging.INFO,
+        stream=sys.stdout,
+    )
+    
+    station_dict = build_station_dict(args.config_folder)
+
+    l3m = xr.Dataset()
+    for stid in  station_dict[args.site]:
+        logger.info(stid)
+        
+        is_promice = False
+        is_gcnet = False
+        filepath = os.path.join(args.folder_l3, stid, stid+'_hour.nc')
+        if os.path.isfile(filepath):
+            is_promice = True
+        else:
+            filepath = os.path.join(args.folder_gcnet, stid+'.csv')
+            if os.path.isfile(filepath):
+                is_gcnet = True
+        if not is_promice and not is_gcnet:            
+            logger.info(stid+' not found either in '+args.folder_l3+' or '+args.folder_gcnet)
+            continue
+
+        l3, _ = loadArr(filepath)
+        
+        if is_gcnet:
+            l3 = gcnet_postprocessing(l3)
+        
+        # lat, lon and alt should be just variables, not coordinates
+        if 'lat' in l3.keys():
+            l3 = l3.reset_coords(['lat', 'lon', 'alt'])
+
+        if len(l3m)==0:
+            # saving attributes of station under an attribute called $stid
+            l3m.attrs[stid] = l3.attrs.copy()
+            # then stripping attributes
+            attrs_list = list(l3.attrs.keys())
+            for k in attrs_list:
+                del l3.attrs[k]
+                
+            l3m = l3.copy()
+        else:
+            # if l3 (older data) is missing variables compared to l3m (newer data)
+            # , then we fill them with nan
+            for v in l3m.data_vars:
```

Comment:
Isn't it possible to get the set of output variables from `variables.csv`, `getColNames` since the output is also a L3 dataset.

Reply:
`xr.concat` is returning a union of all the variables in the input datasets, setting non-present variables to NaN.

Reply:
> `xr.concat` is returning a union of all the variables in the input datasets, setting non-present variables to NaN.

ok, I removed this 
```
                if  v not in l3.data_vars:
                    l3[v] = l3.t_u*np.nan
```

Reply:
> Isn't it possible to get the set of output variables from `variables.csv`, `getColNames` since the output is also a L3 dataset.

I'm not sure what you mean. getColNames would not work on historical GC-Net data (no level attribute...). But the latest station used at a given site gives a good template for which variable should be in the dataset. But I already see some tricky situations:

- at CEN, there is CEN1 that has a `z_stake` and CEN2 that has no `z_stake` (because updated to two-level mast). In that case we want to keep `z_stake` (and therefore the exception in the lines below)

- Still unsolved: at the historical GC-Net stations, the geometry of the stations are changing through time so instead of having instrument height fixed relative to `z_boom` they have a time-dependent `z_T` `z_RH` `z_WS`. Right now those variables are dropped because not present int the more recent GEUS station data. Howver, they will be automatically added as soon as we have an equivalent variable being produced at the L2 or L3_station data.

But your suggestion of passing a `station_info` dict containing all the metadata from a given stations will be helpful because we can add in the future variables to keep or to skip or other station-specific processing before merging.

Reply:
First of all. I'll resolve this conversation because I think to code is working as expected.

# For inspiration
We do know the processing level when invoking `join_l3`. I therefore thought you could use `getColNames` to get a gros list of the columns that are allowed in the output dataset. This will of course require a modification or an alternative function to allow level as parameter instead of using the dataset attributes.

The snippet below shows an example of how I though the columns filtering could be implemented. Note: It is just ment as an example and does not address the station configuration and attributes updates.

```python
def join(input_datasets: List[xr.Dataset], target_columns: Iterable[str]) -> xr.Dataset:
    """
    Join datasets along the time dimension and filter columns.
    
    The newest datasets will be prioritized in case of overlapping time ranges.
    Parameters
    ----------
    input_datasets : List[xr.Dataset]
        List of datasets to join.
    target_columns : List[str]
        List of columns to keep in the final dataset
    """

    # Sort datasets by time in descending order
    input_datasets = sorted(input_datasets, key=lambda ds: ds.time[0], reverse=True)

    datasets = []
    last_start = None
    for ds in input_datasets:
        ds_start = ds.time[0]
        columns = [c for c in target_columns if c in ds]
        datasets.append(
            # Filter columns to align with output level
            ds[columns]
            # Crop time to match previous dataset
            .sel(time=slice(ds_start, last_start))
        )
        last_start = ds_start

    merged_dataset = xr.concat(datasets, dim="time")
    # TODO: Update and maintain station attributes accordingly
    return merged_dataset


target_columns = getColNames(aws.vars, level="L1")
merged_dataset = join(aws.L1, target_columns)
```